omniscidb/QueryEngine/Execute.cpp

```diff
@@ -2787,51 +2666,61 @@ std::vector<std::unique_ptr<ExecutionKernel>> Executor::createHeterogeneousKerne
       data_mgr_->getMemoryInfo(Data_Namespace::MemoryLevel::GPU_LEVEL),
       eo.gpu_input_mem_limit_percent,
       eo.outer_fragment_indices);
-
   CHECK(!ra_exe_unit.input_descs.empty());
 
-  fragment_descriptor.buildFragmentKernelMap(ra_exe_unit,
-                                             shared_context.getFragOffsets(),
-                                             policy,
-                                             available_cpus + available_gpus.size(),
-                                             false, /*multifrag policy unsupported yet*/
-                                             this,
-                                             co.codegen_traits_desc);
-
-  if (allow_single_frag_table_opt && query_mem_descs.count(ExecutorDeviceType::GPU) &&
-      (query_mem_descs.at(ExecutorDeviceType::GPU)->getQueryDescriptionType() ==
-       QueryDescriptionType::Projection) &&
-      table_infos.size() == 1) {
-    const auto max_frag_size = table_infos.front().info.getFragmentNumTuplesUpperBound();
-    if (max_frag_size < query_mem_descs.at(ExecutorDeviceType::GPU)->getEntryCount()) {
-      LOG(INFO) << "Lowering scan limit from "
-                << query_mem_descs.at(ExecutorDeviceType::GPU)->getEntryCount()
-                << " to match max fragment size " << max_frag_size
-                << " for kernel per fragment execution path.";
-      throw CompilationRetryNewScanLimit(max_frag_size);
+  fragment_descriptor.buildFragmentKernelMap(
+      ra_exe_unit, shared_context.getFragOffsets(), policy, this, co.codegen_traits_desc);
+
+  if (!config_->exec.heterogeneous.enable_heterogeneous_execution && eo.with_watchdog &&
+      fragment_descriptor.shouldCheckWorkUnitWatchdog()) {
+    checkWorkUnitWatchdog(ra_exe_unit,
+                          table_infos,
+                          *schema_provider_,
+                          query_comp_descs.begin()->first,
+                          device_count);
+  }
+
+  for (const auto& dt_query_desc : query_mem_descs) {
+    if (policy->getExecutionMode(dt_query_desc.first) ==
+        ExecutorDispatchMode::KernelPerFragment) {
+      VLOG(1) << "Creating one execution kernel per fragment";
```

Comment:
Looks like the log is out of place. Mb move it to where kernels are actually created.